File: src/main/java/no/fintlabs/consumer/resource/ResourceController.java
package no.fintlabs.consumer.resource;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.fint.model.resource.FintResource;
import no.fintlabs.adapter.models.event.RequestFintEvent;
import no.fintlabs.adapter.operation.OperationType;
import no.fintlabs.cache.CacheService;
import no.fintlabs.consumer.kafka.event.EventProducer;
import no.fintlabs.consumer.resource.aspect.IdFieldCheck;
import no.fintlabs.consumer.resource.aspect.WriteableResource;
import no.fintlabs.consumer.resource.event.EventStatusService;
import no.fintlabs.model.resource.FintResources;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.Map;

import static no.fintlabs.consumer.config.EndpointsConstants.*;

@RestController
@RequestMapping("/{resource}")
@RequiredArgsConstructor
@Slf4j
public class ResourceController {

    private final ResourceService resourceService;
    private final CacheService cacheService;
    private final EventProducer eventProducer;
    private final EventStatusService eventStatusService;

    @GetMapping
    public FintResources getResource(
            @PathVariable String resource,
            @RequestParam(defaultValue = "0") int size,
            @RequestParam(defaultValue = "0") int offset,
            @RequestParam(defaultValue = "0") long sinceTimeStamp,
            @RequestParam(required = false) String $filter
    ) {
        return resourceService.getResources(resource, size, offset, sinceTimeStamp, $filter);
    }

    @PostMapping("/$query")
    public FintResources getResourceByOdataFilter(
            @PathVariable String resource,
            @RequestParam(defaultValue = "0") int size,
            @RequestParam(defaultValue = "0") int offset,
            @RequestParam(defaultValue = "0") long sinceTimeStamp,
            @RequestBody(required = false) String $filter
    ) {
        return getResource(resource, size, offset, sinceTimeStamp, $filter);
    }

    @IdFieldCheck
    @GetMapping(BY_ID)
    public FintResource getResourceById(
            @PathVariable String resource,
            @PathVariable String idField,
            @PathVariable String idValue
    ) {
        FintResource fintResource = resourceService.getResourceById(resource, idField, idValue);
        if (fintResource == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
        return fintResource;
    }

    @GetMapping(LAST_UPDATED)
    public Map<String, Long> getLastUpdated(@PathVariable String resource) {
        return Map.of("lastUpdated", cacheService.getCache(resource).getLastUpdated());
    }

    @GetMapping(CACHE_SIZE)
    public Map<String, Integer> getResourceCacheSize(@PathVariable String resource) {
        return Map.of("size", cacheService.getCache(resource).getSize());
    }

    @WriteableResource
    @GetMapping(STATUS_ID)
    public ResponseEntity<Object> getStatus(@PathVariable String resource, @PathVariable String corrId) {
        return eventStatusService.getStatusResponse(resource, corrId);
    }

    @WriteableResource
    @PostMapping
    public ResponseEntity<Void> postResource(
            @PathVariable String resource,
            @RequestBody Object resourceData,
            @RequestParam(name = "validate", required = false) boolean validate
    ) {
        RequestFintEvent requestFintEvent = eventProducer.sendEvent(resource, resourceData, validate ? OperationType.VALIDATE : OperationType.CREATE);
        return ResponseEntity.accepted().header(HttpHeaders.LOCATION, eventStatusService.getStatusHref(requestFintEvent)).build();
    }

    @IdFieldCheck
    @WriteableResource
    @PutMapping(BY_ID)
    public ResponseEntity<Void> putResource(
            @PathVariable String resource,
            @PathVariable String idField,
            @PathVariable String idValue,
            @RequestBody Object resourceData
    ) {
        RequestFintEvent requestFintEvent = eventProducer.sendEvent(resource, resourceData, OperationType.UPDATE);
        return ResponseEntity.accepted().header(HttpHeaders.LOCATION, eventStatusService.getStatusHref(requestFintEvent)).build();
    }

}

File: src/main/java/no/fintlabs/consumer/resource/ResourceController.kt
package no.fintlabs.consumer.resource

import no.fintlabs.adapter.models.event.RequestFintEvent
import no.fintlabs.adapter.operation.OperationType
import no.fintlabs.consumer.config.ConsumerConfiguration
import no.fintlabs.consumer.config.EndpointsConstants
import no.fintlabs.consumer.kafka.event.RequestFintEventProducer
import no.fintlabs.consumer.resource.aspect.IdFieldCheck
import no.fintlabs.consumer.resource.aspect.WriteableResource
import no.fintlabs.consumer.resource.dto.LastUpdatedResponse
import no.fintlabs.consumer.resource.dto.ResourceCacheSizeResponse
import no.fintlabs.consumer.resource.event.*
import no.fintlabs.model.resource.FintResources
import no.novari.fint.model.resource.FintResource
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@RequestMapping("/{resource}")
class ResourceController(
    private val resourceService: ResourceService,
    private val requestFintEventProducer: RequestFintEventProducer,
    private val requestStatusService: RequestStatusService,
    private val consumerConfig: ConsumerConfiguration,
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    @GetMapping
    fun getResource(
        @PathVariable resource: String,
        @RequestParam(defaultValue = "0") size: Int,
        @RequestParam(defaultValue = "0") offset: Int,
        @RequestParam(defaultValue = "0") sinceTimeStamp: Long,
        @RequestParam(required = false, name = "\$filter") filter: String?,
    ): FintResources? =
        resourceService.getResources(
            resource,
            size,
            offset,
            sinceTimeStamp,
            filter,
        )

    @PostMapping("/\$query")
    fun getResourceByOdataFilter(
        @PathVariable resource: String,
        @RequestParam(defaultValue = "0") size: Int,
        @RequestParam(defaultValue = "0") offset: Int,
        @RequestParam(defaultValue = "0") sinceTimeStamp: Long,
        @RequestBody(required = false) filter: String?,
    ): FintResources? = getResource(resource, size, offset, sinceTimeStamp, filter)

    @IdFieldCheck
    @GetMapping(EndpointsConstants.BY_ID)
    fun getResourceById(
        @PathVariable resource: String,
        @PathVariable idField: String?,
        @PathVariable idValue: String,
    ): ResponseEntity<FintResource?> =
        resourceService
            .getResourceById(resource, idField, idValue)
            ?.let { ResponseEntity.ok(it) }
            ?: ResponseEntity.notFound().build()

    @GetMapping(EndpointsConstants.LAST_UPDATED)
    fun getLastUpdated(
        @PathVariable resource: String,
    ): ResponseEntity<LastUpdatedResponse> =
        resourceService.getLastUpdated(resource).let {
            ResponseEntity.ok(LastUpdatedResponse(it))
        }

    @GetMapping(EndpointsConstants.CACHE_SIZE)
    fun getResourceCacheSize(
        @PathVariable resource: String,
    ): ResponseEntity<ResourceCacheSizeResponse> =
        resourceService.getCacheSize(resource).let {
            ResponseEntity.ok(ResourceCacheSizeResponse(it))
        }

    @WriteableResource
    @GetMapping(EndpointsConstants.STATUS_ID)
    fun getStatus(
        @PathVariable resource: String,
        @PathVariable corrId: String,
    ): ResponseEntity<Any?> =
        requestStatusService.getStatusResponse(resource, corrId).let { result ->
            logger.debug("Status of Event: {} returned: {}", corrId, result)
            return when (result) {
                is ResourceCreated -> ResponseEntity.created(result.location).body(result.body)
                is RequestValidated -> ResponseEntity.ok(result.body)
                is ResourceDeleted -> ResponseEntity.noContent().build()
                is RequestAccepted -> ResponseEntity.accepted().build()
                is RequestGone -> ResponseEntity.status(HttpStatus.GONE).build()
                is RequestFailed -> ResponseEntity.status(result.failureType.toHttpStatus()).body(result.body)
            }
        }

    @WriteableResource
    @PostMapping
    fun postResource(
        @PathVariable resource: String,
        @RequestBody resourceData: Any,
        @RequestParam(name = "validate", required = false) validate: Boolean,
    ): ResponseEntity<Nothing> =
        requestFintEventProducer
            .sendEvent(resource.lowercase(), resourceData, validate.getOperationType())
            .toAcceptedResponse()

    @IdFieldCheck
    @WriteableResource
    @PutMapping(EndpointsConstants.BY_ID)
    fun putResource(
        @PathVariable resource: String,
        @PathVariable idField: String,
        @PathVariable idValue: String,
        @RequestBody resourceData: Any,
    ): ResponseEntity<Nothing> =
        requestFintEventProducer
            .sendEvent(resource.lowercase(), resourceData, OperationType.UPDATE)
            .toAcceptedResponse()

    private fun RequestFailed.FailureType.toHttpStatus() =
        when (this) {
            RequestFailed.FailureType.REJECTED -> HttpStatus.BAD_REQUEST
            RequestFailed.FailureType.CONFLICT -> HttpStatus.CONFLICT
            RequestFailed.FailureType.ERROR -> HttpStatus.INTERNAL_SERVER_ERROR
        }

    private fun Boolean.getOperationType() = if (this) OperationType.VALIDATE else OperationType.CREATE

    private fun RequestFintEvent.toLocationUri(): URI = URI.create("${consumerConfig.componentUrl}/$resourceName/status/$corrId")

    private fun RequestFintEvent.toAcceptedResponse(): ResponseEntity<Nothing> = ResponseEntity.accepted().location(toLocationUri()).build()
}
